feat: consolidate dendron configs

hikchoi opened the pull request:
# feat: consolidate dendron configs

This is a draft PR to showcase how the configurations will be organized.

## Overview


- `packages/common-all/src/types/configsdendronConfigs.ts` is where the top most level of configuration is defined.
  - As we go narrower with our namespace, we will define each namespace in a similar fashion.

- I have split up the code so that all each namespace is modularized.
  - Each namespace will define the following:
    - A type that holds all config of this namespace
      - These will be used internally, and in `dendron.yml`
    - A constant object that holds labels and descriptions of every instance of configs in this namespace
      - These are there to provide user friendly descriptions of the configs and will be used in the configuration UI
      - The value, label, and description is bundled up in the `DendronConfigEntry` type.
    - A method that returns a default config of this namespace
      - Each namespace defines this and the parent namespace will simply import and call it
    - Any other useful enum / type definitions related to the configs in the namespace

- Using `packages/common-all/src/types/configs/commands/lookup.ts` to illustrate the above points:
  - Everything here is related to configuring the lookup _command_, so it's under `DendronCommandConfig`.
  - What is defined in `LookupConfig` is what ends up in `dendron.yml`
  - Currently we let the users select what selection behavior should be for note lookup, and the possible choices are defined in `NoteLookupSelectionBehaviorEnum`, and in turn `NoteLookupSelectionBehavior` (which is generated from the enum's keys).
  - Constants in `SELECTION_BEHAVIORS` map the keys in `NoteLookupSelectionBehavior` to respective key's `DendronConfigEntry`s.
    - These will be used to generate useful messages about the lookup command configs in the configuration UI
    - If you introduce new behavior in `NoteLookupSelectionBehaviorEnum`, the compiler will catch this and warn you that you have to add a new label/description here as well. 
      - because `SELECTION_BEHAVIORS` has a type of 
        ```js
          { [key in NoteLookupSelectionBehavior]: DendronConfigEntry<string>}
        ```
  - `genDefaultLookupConfig` will define what the default `LookupConfig` should look like.
  - This will be called by it parent: `genDefaultCommandConfig`.

---

# Pull Request Checklist

You can go to [dendron pull requests](https://wiki.dendron.so/notes/adc39825-77a6-46cf-9c49-2642fcb4248e.html) to see full details for items in this checklist.

## General

### Quality Assurance
- [ ] add a [test](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#writing-tests) for the new feature
- [ ] make sure all the existing [tests](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#running-all-tests) pass
- [ ] do a spot check by running your feature with our [test workspace](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#test-workspace)
- [ ] after you submit your pull request, check the output of our [integration test](https://github.com/dendronhq/dendron/actions) and make sure all tests pass
  - NOTE: if you running mac/linux, check the windows output and vice versa if you are developing on windows

#### Special Cases
- [ ] if your tests changes an existing snaphot, make sure that snapshots are [updated](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#updating-test-snapshots)
- [ ] if you are adding a new language feature (graphically visible in vscode/preview/publishing), make sure that it is included in [test-workspace](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#test-workspace). We use this to manually inspect new changes and for auto regression testiing 

### Docs
- [ ] Make sure that the PR title follows our [commit style](https://wiki.dendron.so/notes/adc39825-77a6-46cf-9c49-2642fcb4248e.html#commit-style)
- [ ] Please summarize the feature or impact in 1-2 lines in the PR description
- [ ] If your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR

Example PR Description
```markdown
# feat: capitalize all foos

This changes capitalizes all occurences of `foo` to `Foo` 

Docs PR: <URL_TO_DOCS_PR>
```

## Special Cases

### First Time PR
- [ ] sign the [CLA](https://en.wikipedia.org/wiki/Contributor_License_Agreement) which will be prompted by our github bot after you submit the PR
- [ ] add your [discord](https://discord.gg/AE3NRw9) alias in the review so that we can give you the [horticulturalist](https://wiki.dendron.so/notes/7c00d606-7b75-4d28-b563-d75f33f8e0d7.html#horticulturalist) badge in our community



### Analytics
- [ ] if you are adding analytics related changes, make sure the [Telemetry](https://wiki.dendron.so/notes/84df871b-9442-42fd-b4c3-0024e35b5f3c.html) docs are updated

kevinslin commented on packages/common-all/src/types/configs/commands/commands.ts:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Constants holding all command config related {@link DendronConfigEntry}
+ */
+export const COMMANDS = {
```
thoughts between using constants vs enums? You can use `Object.values` to get all enum values

hikchoi replied:
The values in the object are also constants that themselves holding other constants or functions that produce constants.

Are you suggesting we have nested enums (not sure if that's a thing)? or we flatten it out and have a giant enum that holds all possible labels and descriptions?

kevinslin commented on packages/common-all/src/types/configs/workspace/journal.ts:
```diff
@@ -0,0 +1,126 @@
+ * @param value {@link dayOfWeekNumber}
+ * @returns DendronConfigEntry
+ */
+const FIRST_DAY_OF_WEEK = (
```
just a heads up that we're currently not supporting this option due to limitation in calendar widget. we'll add support back in eventually but FYI

hikchoi replied:
I am aware of this. I also have been taking note on other configs with some peculiarities that we need to address once we start actually migrating.

kevinslin reviewed (commented):
This looks great. Left some minor comments but go ahead with this. For this commit, is the plan to carry out [phase I](https://wiki.dendron.so/notes/5mfPAJFfNFrD16RbGOs9h.html#phase-i) of config migration? Can we have this ready to be merged for next week's release?

For deprecation notice, since we'll do automatic migration, we might not need to do any sort of deprecation notice besides announcing it in our release notes. FOr migrating to the new configuration, can we do it piecemeal?

Eg. Week1 -> migrate command config. Week2 -> migrate 2 more sections. Week3 -> migrate everything...

hikchoi commented:
> This looks great. Left some minor comments but go ahead with this. For this commit, is the plan to carry out [phase I](https://wiki.dendron.so/notes/5mfPAJFfNFrD16RbGOs9h.html#phase-i) of config migration? Can we have this ready to be merged for next week's release?
> 
> For deprecation notice, since we'll do automatic migration, we might not need to do any sort of deprecation notice besides announcing it in our release notes. FOr migrating to the new configuration, can we do it piecemeal?
> 
> Eg. Week1 -> migrate command config. Week2 -> migrate 2 more sections. Week3 -> migrate everything...


Yeah, I think it makes sense to announce what's going to change and then slowly migrate the configurations to the new one.
There are a lot of things to fix (splitting the effect of config so that the dupes in multiple areas are respected) so a piecemeal migration is likely better than one giant one.

I'll get it ready by next release, and also write up an announcement to add to the release notes of what the changes will entail.

hikchoi commented:
UPDATE: 

Please check the overview in the main body of this PR. Since this PR is waaaay too big, I go over the general structure of what is being added.

This PR does not change any existing behavior of the config. It just adds the newly organized config modules.
Over the next few weeks I will incrementally migrate the new configs in chunks.
Due to potential package namespace collision issues, I have not exported any of the newly added modules yet; they will be exported from `@dendronhq/common-all` as I actually migrate the configs.

I will also be adding a document that goes over exactly which config is going to be mapped to which in a separate docs PR, which can be added to the upcoming release notes.

hikchoi commented:
Here is an exhaustive list of changes to every config key

- This can also be found in [[Diff of All Changes|user.hikchoi.scratch.2021.09.14.diff-of-all-changes]]
- A _user-friendly_ version can be found in [[User Friendly|user.hikchoi.scratch.2021.09.14.diff-of-all-changes.user-friendly]]

## Dendron Configurations

- noCaching?: boolean;
  - Renamed to `enableCaching`
  - Now a required property
  - Now explicitly defaults to `true`
  - New location: `DendronConfig.workspace.enableCaching`
- maxPreviewsCached?: number;
  - Now a required property
  - Now explicitly defaults to the value `10`
  - New location: `DendronConfig.workspace.maxPreviewsCached`
- noTelemetry?: boolean;
  - Renamed to `enableTelemetry`
  - Now a required property
  - Now explicitly defaults to `true`
  - New location: `DendronConfig.workspace.enableTelemetry`
- version: number;
  - deprecated
- dendronVersion?: string;
  - New location: `DendronConfig.workspace.dendronVersion`
- site: DendronSiteConfig;
  - See `Site Configurations` below
- lookup: LookupConfig;
  - New location: `DendronConfig.commands.lookup`
  - `lookup.note.selectionType` renamed to `lookup.note.selectionMode`
  - enum values of selection modes are renamed
    - `selectionExtract` renamed to `extract`
    - `selection2link` renamed to `link`
- journal: JournalConfig;
  - New location: `DendronConfig.workspace.journal`
  - `journal.firstDayOfWeek` temporarily removed
- scratch?: ScratchConfig;
  - Now a required property
  - New location: `DendronConfig.workspace.scratch`
- insertNoteLink?: InsertNoteLinkConfig;
  - Now a required property
  - `aliasMode` now explicitly defaults to `none`
  - `enableMultiSelect` now explicitly defaults to `false`
  - New location: `DendronConfig.commands.insertNoteLink`
  - `insertNoteLinkConfig.multiselect` renamed to `insertNoteLink.enableMultiSelect`
- workspaces?: { [key: string]: DWorkspaceEntry | undefined };
  - New location: `DendronConfig.workspace.workspaces`
- seeds?: { [key: string]: SeedEntry | undefined };
  - New location: `DendronConfig.workspace.seeds`
- vaults: DVault[];
  - New location: `DendronConfig.workspace.vaults`
- hooks?: DHookDict;
  - New location: `DendronConfig.workspace.hooks`
- lookupConfirmVaultOnCreate?: boolean;
  - Now part of `lookupConfig`
  - Renamed to `confirmVaultOnCreate`
  - Now a required property
  - Now explicitly defaults to `false`
  - New location: `DendronConfig.commands.lookup.note.confirmVaultOnCreate`
- useFMTitle?: boolean;
  - Renamed to `enableFMTitle`
  - Now a required property in `global` namespace.
  - Now explicitly set to `true` in `global` namespace.
  - Remains optional in `preview` and `publishing`
  - When explicitly set in `preview` or `publishing` namespace, it will have precedence over the value set in `global` namespace.
  - Now duplicated across `global`, `preview`, and `publising` namespace
- useNoteTitleForLink?: boolean;
  - Renamed to `enableNoteTitleForLink`
  - Now a required property in `global` namespace.
  - Now explicitly set to `true` in `global` namespace.
  - Remains optional in `preview` and `publishing`
  - When explicitly set in `preview` or `publishing` namespace, it will have precedence over the value set in `global` namespace.
  - Now duplicated across `global`, `preview`, and `publising` namespace 
- mermaid?: boolean;
  - Renamed to `enableMermaid`
  - Now a required property in `global` namespace.
  - Now explicitly set to `true` in `global` namespace.
  - Remains optional in `preview` and `publishing`
  - When explicitly set in `preview` or `publishing` namespace, it will have precedence over the value set in `global` namespace.
  - Now duplicated across `global`, `preview`, and `publising` namespace 
- useNunjucks?: boolean;
  - Renamed to `enableNunjucks`
  - Now a required property in `global` namespace.
  - Now explicitly set to `false` in `global` namespace.
  - Remains optional in `preview` and `publishing`
  - When explicitly set in `preview` or `publishing` namespace, it will have precedence over the value set in `global` namespace.
  - Now duplicated across `global`, `preview`, and `publising` namespace 
- usePrettyRefs?: boolean;
  - Renamed to `enablePrettyRefs`
  - Now a required property in `global` namespace.
  - Now explicitly set to `true` in `global` namespace.
  - Remains optional in `preview` and `publishing`
  - When explicitly set in `preview` or `publishing` namespace, it will have precedence over the value set in `global` namespace.
  - Now duplicated across `global`, `preview`, and `publising` namespace 
- useKatex?: boolean;
  - Renamed to `enableKatex`
  - Now a required property in `global` namespace.
  - Now explicitly set to `true` in `global` namespace.
  - Remains optional in `preview` and `publishing`
  - When explicitly set in `preview` or `publishing` namespace, it will have precedence over the value set in `global` namespace.
  - Now duplicated across `global`, `preview`, and `publising` namespace 
- hierarchyDisplay?: boolean;
  - Renamed to `enabledHierarchyDisplay`
  - Now a required property in `global` namespace.
  - Now explicitly set to `true` in `global` namespace.
  - Remains optional in `preview` and `publishing`
  - When explicitly set in `preview` or `publishing` namespace, it will have precedence over the value set in `global` namespace.
  - Now duplicated across `global`, `preview`, and `publising` namespace 
- hierarchyDisplayTitle?: string;
  - Now explicitly set to `children` only in `global` namespace, but remains optional.
  - When explicitly set in `preview` or `publishing` namespace, it will have precedence over the value set in `global` namespace.
  - Now duplicated across `global`, `preview`, and `publising` namespace 
- graph?: DendronGraphConfig;
  - Now a required property
  - `zoomSpeed` explicitly defaults to the value `1`
  - New location: `DendronConfig.workspace.graph`
- noAutoCreateOnDefinition?: boolean;
  - Renamed to `enableAutoCreateOnDefinition`
  - Now a require property
  - Now explicitly defaults to `false`
  - New location: `DendronConfig.workspace.enableAutoCreateOnDefinition`
- noLegacyNoteRef?: boolean;
  - Renamed to `enableLegacyNoteRef`
  - Now a required property in `global` namespace.
  - Now explicitly set to `false` in `global` namespace.
  - Remains optional in `preview` and `publishing`
  - When explicitly set in `preview` or `publishing` namespace, it will have precedence over the value set in `global` namespace.
  - Now duplicated across `global`, `preview`, and `publising` namespace 
- noXVaultWikiLink?: boolean;
  - Renamed to `enableXVaultWikiLink`
  - Now a required property
  - Now explicitly defaults to `false`
  - New location: `DendronConfig.workspace.enableXVaultWikiLink`
- initializeRemoteVaults?: boolean;
  - Now a required property
  - Now explicitly defaults to `true`
  - New location: `DendronConfig.workspace.initializeRemoteVault`
- feedback?: boolean;
  - New location: `DendronConfig.workspace.feedback`
- apiEndpoint?: string;
  - New location: `DendronConfig.workspace.apiEndpoint`
- defaultInsertHierarchy?: string;
  - Now part of `insertNoteConfig`
  - renamed to `initialValue`
  - Explicitly defaults to `templates` but remains optional
  - New location: `DendronConfig.commands.insertNote.initialValue`
- dev?: DendronDevConfig;
  - New location: `DendronConfig.dev`
- workspaceVaultSync?: DVaultSync;
  - Renamed to `workspaceVaultSyncMode`
  - Now a required property
  - Now explicitly defaults to `noCommit`
  - New location: `DendronConfig.workspace.workspaceVaultSyncMode`
- randomNote?: RandomNoteConfig;
  - Now a required property
  - Properties `randomNote.include` and `randomNote.exclude` remains optional
  - New location: `DendronConfig.commands.randomNote`
- autoFoldFrontmatter?: boolean;
  - Renamed to `enableAutoFoldFrontmatter`
  - Now a required property
  - Now explicitly defaults to `true`
  - New location: `DendronConfig.workspace.enableAutoFoldFrontmatter`
- insertNoteIndex?: InsertNoteIndexConfig;
  - Now a required property
  - New location: `DendronConfig.commands.insertNoteIndex`
  - `insertNoteIndexConfig.marker` renamed to `insertNoteIndexConfig.enableMarker`
  - `insertNoteIndexConfig.enableMarker` is now a required property
  - `insertNoteIndexConfig.enableMarker` now explicitly defaults to `false`
- maxNoteLength?: number;
  - Now a required property
  - Now explicitly defaults to the value `204800`
  - New location: `DendronConfig.workspace.maxNoteLength`

## Site Configurations
- assetsPrefix?: string;
  - New location: `DendronConfig.publishing.assetsPrefix`
- canonicalBaseUrl?: string;
  - New location: `DendronConfig.publishing.canonicalBaseUrl`
- copyAssets?: boolean;
  - Now a required property
  - Now explicitly defaults to `true`
  - New location: `DendronConfig.publishing.copyAssets`
- customHeaderPath?: string;
  - New location: `DendronConfig.publishing.customHeaderPath`
- ga_tracking?: string;
  - Now part of `GoogleAnalyticsConfig`
  - New location: `DendronConfig.publishing.ga.tracking`
- siteFaviconPath?: string;
  - New location: `DendronConfig.publishing.siteFaviconPath`
- logo?: string;
  - Renamed to `logoPath`
  - New location: `DendronConfig.publishing.logoPath`
- siteIndex?: string;
  - New location: `DendronConfig.publishing.siteIndex`
- siteHierarchies: string[];
  - New location: `DendronConfig.publishing.siteHierarchies`
- siteLastModified?: boolean;
  - Renamed to `enableSiteLastModified`
  - Now a required property
  - Now explicitly defaults to `true`
  - New location: `DendronConfig.publishing.enableSiteLastModified`
- siteRootDir: string;
  - New location: `DendronConfig.publishing.siteRootDir`
- siteRepoDir?: string;
  - New location: `DendronConfig.publishing.siteRepoDir`
- siteNotesDir?: string;
  - deprecated
- siteUrl?: string;
  - New location: `DendronConfig.publishing.siteUrl`
- githubCname?: string;
  - Now part of `GithubConfig`
  - Renamed to `cname`
  - New location: `DendronConfig.publishing.github.cname`
- gh_edit_link?: string;
  - Now part of `GithubConfig`
  - Renamed to `enableEditLink`
  - Now a required property
  - Now explicitly defaults to `true`
  - New location: `DendronConfig.publishing.github.enableEditLink`
- gh_edit_link_text?: string;
  - Now part of `GithubConfig`
  - Renamed to `editLinkText`
  - Remains optional but explicitly defaults to `Edit this page on GitHub`
  - New location: `DendronConfig.publishing.github.editLinkText`
- gh_edit_branch?: string;
  - Now part of `GithubConfig`
  - Renamed to `editBranch`
  - Remains optional but explicitly defaults to `main`
  - New location: `DendronConfig.publishing.github.editBranch`
- gh_edit_view_mode?: "tree" | "edit";
  - Now part of `GithubConfig`
  - Renamed to `editViewMode`
  - Remains optional but explicitly defaults to `tree`
  - New location: `DendronConfig.publishing.github.editViewMode`
- gh_edit_repository?: string;
  - Now part of `GithubConfig`
  - Renamed to `editRepository`
  - New location: `DendronConfig.publishing.github.editRepository`
- usePrettyRefs?: boolean;
  - Renamed to `enablePrettyRefs`
  - Now a required property in `global` namespace.
  - Now explicitly set to `true` in `global` namespace.
  - Remains optional in `preview` and `publishing`
  - When explicitly set in `preview` or `publishing` namespace, it will have precedence over the value set in `global` namespace. 
  - Now duplicated in `global`, `preview`, and `publishing`
- hideBlockAnchors?: boolean;
  - deprecated
- showFrontMatterTags?: boolean;
  - Renamed to `enableFrontmatterTags`
  - Now a required property
  - Now explicitly defaults to `true`
  - New location: `DendronConfig.publishing.enableFrontmatterTags`
- noRandomlyColoredTags?: boolean;
  - Renamed to `enableRandomlyColoredTags`
  - Now a required property
  - Now explicitly defaults to `true`
  - New location: `DendronConfig.publishing.enableRandomlyColoredTags`
- config?: { [key: string]: HierarchyConfig };
  - Renamed to `hierarchy`
  - New location: `DendronConfig.publishing.hierarchy`
- duplicateNoteBehavior?: DuplicateNoteBehavior;
  - New location: `DendronConfig.publishing.duplicateNoteBehavior`
- writeStubs?: boolean;
  - Now a required property
  - Now explicitly defaults to `false`
  - New location: `DendronConfig.publishing.writeStubs`
- title?: string;
  - Now part of `SEOConfig`
  - New location: `DendronConfig.publishing.seo.title`
- description?: string;
  - Now part of `SEOConfig`
  - New location: `DendronConfig.publishing.seo.description`
- author?: string;
  - Now part of `SEOConfig`
  - New location: `DendronConfig.publishing.seo.author`
- twitter?: string;
  - Now part of `SEOConfig`
  - New location: `DendronConfig.publishing.seo.twitter`
- image?: string;
  - Now part of `SEOConfig`
  - New location: `DendronConfig.publishing.seo.image`
- useContainers?: boolean;
  - Renamed to `enableContainers`
  - Now a required property
  - Now explicitly defaults to `false`
  - New location: `DendronConfig.publishing.enableContainers`
- generateChangelog?: boolean;
  - Now a required property
  - Now explicitly defaults to `false`
  - New location: `DendronConfig.publishing.generateChangelog`
- previewPort?: boolean;
  - Type fixed to `number`
  - New location: `DendronConfig.publishing.previewPort`
- segmentKey?: string;
  - New location: `DendronConfig.publishing.segmentKey`
- cognitoUserPoolId?: string;
  - New location: `DendronConfig.publishing.cognitoUserPoolId`
- cognitoClientId?: string;
  - New location: `DendronConfig.publishing.cognitoClientId`
- usePrettyLinks?: boolean;
  - Renamed to `enablePrettyLinks`
  - Now a required property
  - Now explicitly defaults to `true`
  - New location: `DendronConfig.publishing.enablePrettyLinks`

kevinslin commented:
Great job on putting in the ground work. Configuration renames look good, just a few changes suggested below

- noCaching?: boolean;
  - Renamed to `enableCaching`
  - Now a required property
  - Now explicitly defaults to `true`
  - New location: `DendronConfig.workspace.enableCaching`
  > we initially had this because we weren't sure if caching would break things. it hasn't so far - lets just remove this entirely

- noTelemetry?: boolean;
  - Renamed to `enableTelemetry`
  - Now a required property
  - Now explicitly defaults to `true`
  - New location: `DendronConfig.workspace.enableTelemetry`
  > can we change this to `disableTelemetry` and keep it as optional?

- noLegacyNoteRef?: boolean;
  - Renamed to `enableLegacyNoteRef`
  - Now a required property in `global` namespace.
  - Now explicitly set to `false` in `global` namespace.
  - Remains optional in `preview` and `publishing`
  - When explicitly set in `preview` or `publishing` namespace, it will have precedence over the value set in `global` namespace.
  - Now duplicated across `global`, `preview`, and `publising` namespace 
  > we should just disable this altogether and remove the option

- initializeRemoteVaults?: boolean;
  - Now a required property
  - Now explicitly defaults to `true`
  - New location: `DendronConfig.workspace.initializeRemoteVault`
  > change to `enableRemoteVaultInit`

- autoFoldFrontmatter?: boolean;
  - Renamed to `enableAutoFoldFrontmatter`
  - Now a required property
  - Now explicitly defaults to `true`
  - New location: `DendronConfig.workspace.enableAutoFoldFrontmatter`
  > this has actually caused a bunch of issues. can we change the default to false?

kevinslin commented:
> Note that value is defined to be optional, as some configs can have arbitrary values. In which case passing in the correct type T is meaningless, but is there for completeness' sake.

typescript supports initializing generics, would rewrite to the following in that case
```
export type DendronConfigEntry<T=any> = {
  value?: T;
  label: string;
  desc: string;
};
```

kevinslin commented on packages/common-all/src/types/configs/base.ts:
```diff
@@ -0,0 +1,38 @@
+ * For config entries that can be an arbitrary value, only specify the label and description.
+ * For config entries that have pre-defined choices, provide the value as well as label and description specific to that value.
+ */
+export type DendronConfigEntry<T> = {
```
since value is optional we can make `T=any` to cut down on boilerplate

hikchoi replied:
TIL 😅  still so much to learn about ts.

kevinslin reviewed (changes requested):
see config changes in conversation
